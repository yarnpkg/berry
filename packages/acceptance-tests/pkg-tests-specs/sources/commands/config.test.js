const {
  fs: {mkdirp, writeFile, createTemporaryFolder},
} = require(`pkg-tests-core`);

const RC_FILENAME = `.spec-yarnrc`;
const SUBFOLDER = `subfolder`;
const FAKE_REGISTRY_URL = `http://yarn.test.registry`;
const FAKE_LOCAL_APP_DATA = `LOCAL_APP_DATA`;
const FAKE_WORKSPACE_ROOT = `WORKSPACE_ROOT`;
const FAKE_HOME = `HOME`;

const FILTER = new RegExp([
  `initScope`,
  `lastUpdateCheck`,
  `defaultLanguageName`,
].join(`|`));

const environments = {
  [`folder without rcfile in ancestry`]: async () => {
    // Nothing to do
  },
  [`folder with rcfile`]: async ({path}) => {
    await writeFile(`${path}/${SUBFOLDER}/${SUBFOLDER}/${RC_FILENAME}`, `initScope: my-test\n`);
  },
  [`folder with rcfile without trailing newline`]: async ({path}) => {
    await writeFile(`${path}/${SUBFOLDER}/${SUBFOLDER}/${RC_FILENAME}`, `initScope: my-test`);
  },
  [`folder with rcfile and rc in parent`]: async ({path}) => {
    await writeFile(`${path}/${SUBFOLDER}/${SUBFOLDER}/${RC_FILENAME}`, `initScope: my-test\n`);
    await writeFile(`${path}/${SUBFOLDER}/${RC_FILENAME}`, `initScope: value-to-override\nlastUpdateCheck: 1555784893958\n`);
  },
  [`folder with rcfile and rc in ancestor parent`]: async ({path}) => {
    await writeFile(`${path}/${SUBFOLDER}/${SUBFOLDER}/${RC_FILENAME}`, `initScope: my-test\n`);
    await writeFile(`${path}/${RC_FILENAME}`, `initScope: value-to-override\nlastUpdateCheck: 1555784893958\n`);
  },
  [`folder with rcfile and rc in home folder`]: async ({path, homePath}) => {
    await writeFile(`${homePath}/${RC_FILENAME}`, `initScope: value-to-override\ndefaultLanguageName: python\n`);
    await writeFile(`${path}/${RC_FILENAME}`, `initScope: my-test\nlastUpdateCheck: 1555784893958\n`);
  },
};

function cleanupPlainOutput(output, path, homePath) {
  // Replace multiple consecutive spaces with one space.
  // The output of the config command is aligned according to the longest value, which probably
  // contains `path`. In other words, the formatting depends on the length of `path`.
  output = output.replace(/  +/g, ` - `);

  // replace the generated workspace root with a constant
  output = output.replace(new RegExp(path, `g`), FAKE_WORKSPACE_ROOT);

  // replace the generated home folder with a constant
  output = output.replace(new RegExp(homePath, `g`), FAKE_HOME);

  // replace the generated registry server URL with a constant
  output = output.replace(/http:\/\/localhost:\d+/g, FAKE_REGISTRY_URL);

  // replace the default global folder with a constant
  output = output.replace(/[^"]+\/\.?yarn\/berry/ig, FAKE_LOCAL_APP_DATA);

  output = output.split(/\n/).filter(line => {
    return line.match(FILTER) !== null;
  }).join(`\n`);

  return output;
}

function cleanupJsonOutput(output, path, homePath) {
  let outputObject;
  try {
    outputObject = JSON.parse(output);
  } catch (e) {
    return cleanupPlainOutput(output, path, homePath);
  }

  // the default globalFolder contains the user's home folder, override that value
  outputObject.globalFolder.default = `DEFAULT_GLOBAL_FOLDER`;

  // replace the generated registry server URL with a constant
  outputObject.npmRegistryServer.effective = FAKE_REGISTRY_URL;

  const pathRegExp = new RegExp(path, `g`);
  const homePathRegExp = new RegExp(homePath, `g`);

  for (const setting of Object.values(outputObject)) {
    if (typeof setting.source === `string`) {
      setting.source = setting.source.replace(pathRegExp, FAKE_WORKSPACE_ROOT);
      setting.source = setting.source.replace(homePathRegExp, FAKE_HOME);
    }

    if (typeof setting.default === `string`) {
      setting.default = setting.default.replace(pathRegExp, FAKE_WORKSPACE_ROOT);
      setting.default = setting.default.replace(homePathRegExp, FAKE_HOME);
    }

    if (typeof setting.effective === `string`) {
      setting.effective = setting.effective.replace(pathRegExp, FAKE_WORKSPACE_ROOT);
      setting.effective = setting.effective.replace(homePathRegExp, FAKE_HOME);
    }
  }

  return JSON.stringify(outputObject);
}

const options = {
  [`without flags`]: {cleanupStdout: cleanupPlainOutput, flags: []},
  [`showing the source`]: {cleanupStdout: cleanupPlainOutput, flags: [`--why`]},
  [`showing explanation`]: {cleanupStdout: cleanupPlainOutput, flags: [`--verbose`]},
  [`as json`]: {cleanupStdout: cleanupJsonOutput, flags: [`--json`]},
};

describe(`Commands`, () => {
  describe(`config`, () => {
    for (const [environmentDescription, environment] of Object.entries(environments)) {
      for (const [optionDescription, {flags, cleanupStdout}] of Object.entries(options)) {
        test(`test (${environmentDescription} / ${optionDescription})`, makeTemporaryEnv({}, async ({path, run, source}) => {
          const cwd = `${path}/${SUBFOLDER}/${SUBFOLDER}`;
          const homePath = await createTemporaryFolder();

          await mkdirp(cwd);
          await environment({path, homePath});

          let code;
          let stdout;
          let stderr;

          try {
            ({code, stdout, stderr} = await run(`config`, ...flags, {cwd, env: {YARN_RC_FILENAME: RC_FILENAME, HOME: homePath, USERPROFILE: homePath}}));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          stdout = cleanupStdout(stdout, path, homePath);
          stderr = cleanupPlainOutput(stderr, path, homePath);

          expect({code, stdout, stderr}).toMatchSnapshot();
        }));
      }
    }
  });
});
