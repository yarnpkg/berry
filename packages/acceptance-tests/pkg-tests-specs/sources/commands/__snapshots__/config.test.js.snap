// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands config test (folder with rcfile / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"node\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":\\"my-test\\",\\"source\\":\\"WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc\\",\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":null,\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder with rcfile / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'node'
➤ YN0000: initScope - Scope used when creating packages via the init command - 'my-test'
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - null",
}
`;

exports[`Commands config test (folder with rcfile / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - <default> - 'node'
➤ YN0000: initScope - WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc - 'my-test'
➤ YN0000: lastUpdateCheck - <default> - null",
}
`;

exports[`Commands config test (folder with rcfile / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'node'
➤ YN0000: initScope - 'my-test'
➤ YN0000: lastUpdateCheck - null",
}
`;

exports[`Commands config test (folder with rcfile and rc in ancestor parent / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"node\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":\\"my-test\\",\\"source\\":\\"WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc\\",\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":\\"1555784893958\\",\\"source\\":\\"WORKSPACE_ROOT/.spec-yarnrc\\",\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder with rcfile and rc in ancestor parent / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'node'
➤ YN0000: initScope - Scope used when creating packages via the init command - 'my-test'
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in ancestor parent / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - <default> - 'node'
➤ YN0000: initScope - WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc - 'my-test'
➤ YN0000: lastUpdateCheck - WORKSPACE_ROOT/.spec-yarnrc - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in ancestor parent / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'node'
➤ YN0000: initScope - 'my-test'
➤ YN0000: lastUpdateCheck - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in home folder / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"python\\",\\"source\\":\\"HOME/.spec-yarnrc\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":\\"my-test\\",\\"source\\":\\"WORKSPACE_ROOT/.spec-yarnrc\\",\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":\\"1555784893958\\",\\"source\\":\\"WORKSPACE_ROOT/.spec-yarnrc\\",\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder with rcfile and rc in home folder / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'python'
➤ YN0000: initScope - Scope used when creating packages via the init command - 'my-test'
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in home folder / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - HOME/.spec-yarnrc - 'python'
➤ YN0000: initScope - WORKSPACE_ROOT/.spec-yarnrc - 'my-test'
➤ YN0000: lastUpdateCheck - WORKSPACE_ROOT/.spec-yarnrc - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in home folder / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'python'
➤ YN0000: initScope - 'my-test'
➤ YN0000: lastUpdateCheck - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in parent / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"node\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":\\"my-test\\",\\"source\\":\\"WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc\\",\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":\\"1555784893958\\",\\"source\\":\\"WORKSPACE_ROOT/subfolder/.spec-yarnrc\\",\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder with rcfile and rc in parent / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'node'
➤ YN0000: initScope - Scope used when creating packages via the init command - 'my-test'
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in parent / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - <default> - 'node'
➤ YN0000: initScope - WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc - 'my-test'
➤ YN0000: lastUpdateCheck - WORKSPACE_ROOT/subfolder/.spec-yarnrc - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile and rc in parent / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'node'
➤ YN0000: initScope - 'my-test'
➤ YN0000: lastUpdateCheck - '1555784893958'",
}
`;

exports[`Commands config test (folder with rcfile without trailing newline / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"node\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":\\"my-test\\",\\"source\\":\\"WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc\\",\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":null,\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder with rcfile without trailing newline / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'node'
➤ YN0000: initScope - Scope used when creating packages via the init command - 'my-test'
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - null",
}
`;

exports[`Commands config test (folder with rcfile without trailing newline / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - <default> - 'node'
➤ YN0000: initScope - WORKSPACE_ROOT/subfolder/subfolder/.spec-yarnrc - 'my-test'
➤ YN0000: lastUpdateCheck - <default> - null",
}
`;

exports[`Commands config test (folder with rcfile without trailing newline / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'node'
➤ YN0000: initScope - 'my-test'
➤ YN0000: lastUpdateCheck - null",
}
`;

exports[`Commands config test (folder without rcfile in ancestry / as json) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "{\\"key\\":\\"defaultLanguageName\\",\\"effective\\":\\"node\\",\\"description\\":\\"Default language mode that should be used when a package doesn't offer any insight\\",\\"type\\":\\"STRING\\",\\"default\\":\\"node\\"}
{\\"key\\":\\"initScope\\",\\"effective\\":null,\\"description\\":\\"Scope used when creating packages via the init command\\",\\"type\\":\\"STRING\\",\\"default\\":null}
{\\"key\\":\\"lastUpdateCheck\\",\\"effective\\":null,\\"description\\":\\"Last timestamp we checked whether new Yarn versions were available\\",\\"type\\":\\"STRING\\",\\"default\\":null}",
}
`;

exports[`Commands config test (folder without rcfile in ancestry / showing explanation) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - Default language mode that should be used when a package doesn't offer any insight - 'node'
➤ YN0000: initScope - Scope used when creating packages via the init command - null
➤ YN0000: lastUpdateCheck - Last timestamp we checked whether new Yarn versions were available - null",
}
`;

exports[`Commands config test (folder without rcfile in ancestry / showing the source) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - <default> - 'node'
➤ YN0000: initScope - <default> - null
➤ YN0000: lastUpdateCheck - <default> - null",
}
`;

exports[`Commands config test (folder without rcfile in ancestry / without flags) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: defaultLanguageName - 'node'
➤ YN0000: initScope - null
➤ YN0000: lastUpdateCheck - null",
}
`;
